src/main/java/org/opensearch/neuralsearch/highlight/SemanticHighlightingResponseProcessor.java

```diff
@@ -0,0 +1,390 @@
+     */
+    private void applyHighlightsToHit(SearchHit hit, List<Map<String, Object>> highlights, String fieldName) {
+        try {
+            Map<String, Object> source = hit.getSourceAsMap();
+            if (source == null) {
+                return;
+            }
+
+            String text = (String) source.get(fieldName);
+            if (text != null) {
+                String highlightedText = applyPositionHighlights(text, highlights);
+
+                Map<String, HighlightField> highlightFields = Optional.ofNullable(hit.getHighlightFields()).orElse(new HashMap<>());
+
+                HighlightField highlightField = new HighlightField(fieldName, new Text[] { new Text(highlightedText) });
+                highlightFields.put(fieldName, highlightField);
+                hit.highlightFields(highlightFields);
+            }
+        } catch (Exception e) {
+            log.error(String.format(Locale.ROOT, "error applying highlights to hit %s", hit.getId()), e);
+        }
+    }
+
+    /**
+     * Apply position-based highlights to text using ProcessorUtils for validation
+     */
+    private String applyPositionHighlights(String text, List<Map<String, Object>> highlights) {
+        try {
+            List<Map<String, Object>> validHighlights = new ArrayList<>();
+            for (Map<String, Object> highlight : highlights) {
+                Object startObj = highlight.get(START_KEY);
+                Object endObj = highlight.get(END_KEY);
+
+                if (ProcessorUtils.isNumeric(startObj) && ProcessorUtils.isNumeric(endObj)) {
+                    validHighlights.add(highlight);
+                }
+            }
+
+            if (validHighlights.isEmpty()) {
+                return text;
+            }
+
+            StringBuilder result = new StringBuilder(text);
+            for (int i = validHighlights.size() - 1; i >= 0; i--) {
+                Map<String, Object> highlight = validHighlights.get(i);
+                int start = ((Number) highlight.get(START_KEY)).intValue();
+                int end = ((Number) highlight.get(END_KEY)).intValue();
+
+                if (start >= 0 && end <= text.length() && start < end) {
+                    result.insert(end, postTag);
+                    result.insert(start, preTag);
+                }
+            }
+
+            return result.toString();
+
+        } catch (Exception e) {
+            log.error(String.format(Locale.ROOT, "error applying position highlights, returning original text"), e);
+            return text;
+        }
+    }
+
+    /**
+     * Apply batch highlight results to search response using valid hits list
+     */
+    private SearchResponse applyBatchHighlightResults(
+        SearchResponse response,
+        List<List<Map<String, Object>>> batchResults,
+        List<SearchHit> validHits,
+        String semanticHighlightField
+    ) {
+        try {
+            for (int i = 0; i < validHits.size() && i < batchResults.size(); i++) {
```

Comment:
Shouldn't we validate that `validHits.size() == batchResults.size()` and throw an exception otherwise?